backend/editor/controllers/search_controller.py

```diff
@@ -0,0 +1,374 @@
+
+    def build_cypher_query(self, index: int) -> tuple[str, str]:
+        return (
+            "(n)-[:is_child_of]->(:ENTRY {id: $" + get_query_param_name(index) + "})",
+            self.filter_value,
+        )
+
+
+@dataclass
+class AncestorFilterSearchTerm(FilterSearchTerm):
+    filter_value: str
+
+    def build_cypher_query(self, index: int) -> tuple[str, str]:
+        return (
+            "(n)<-[:is_child_of*]-(:ENTRY {id: $" + get_query_param_name(index) + "})",
+            self.filter_value,
+        )
+
+
+@dataclass
+class DescendantFilterSearchTerm(FilterSearchTerm):
+    filter_value: str
+
+    def build_cypher_query(self, index: int) -> tuple[str, str]:
+        return (
+            "(n)-[:is_child_of*]->(:ENTRY {id: $" + get_query_param_name(index) + "})",
+            self.filter_value,
+        )
+
+
+@dataclass
+class Query:
+    project_id: str
+    name_search_terms: list[str]
+    filter_search_terms: list[FilterSearchTerm]
+
+
+def split_query_into_search_terms(query: str) -> list[str]:
+    """
+    Queries should be split by whitespaces that are not inside quotes
+    """
+    query = query.strip()
+    search_terms = []
+
+    inside_quotes = False
+    term_start = 0
+
+    for term_end in range(len(query)):
+        if query[term_end] == '"':
+            inside_quotes = not inside_quotes
+        # If we are not inside quotes and we encounter a whitespace
+        # we are at the end of the current search term
+        elif query[term_end] == " " and not inside_quotes:
+            search_term = query[term_start:term_end]
+            search_terms.append(search_term)
+            term_start = term_end + 1
+
+    search_terms.append(query[term_start:])
+
+    return search_terms
+
+
+def parse_filter_search_term(search_term: str) -> FilterSearchTerm | None:
+    if ":" not in search_term:
+        return None
+
+    filter_name, filter_value = search_term.split(":", maxsplit=1)
+
+    if filter_value.startswith('"') and filter_value.endswith('"'):
+        filter_value = filter_value[1:-1]
+
+    if '"' in filter_value:
+        return None
+
+    match filter_name:
+        case "is":
+            if filter_value == "root":
+                return IsFilterSearchTerm(filter_value)
+        case "language" | "not(language)":
+            if len(filter_value) != 2 and filter_value.isalpha():
+                return None
+
+            return (
+                LanguageFilterSearchTerm(filter_value)
+                if filter_name == "language"
+                else LanguageFilterSearchTerm(filter_value, negated=True)
+            )
+        case "parent":
+            return ParentFilterSearchTerm(filter_value)
+        case "child":
+            return ChildFilterSearchTerm(filter_value)
+        case "ancestor":
+            return AncestorFilterSearchTerm(filter_value)
+        case "descendant":
+            return DescendantFilterSearchTerm(filter_value)
+        case _:
+            return None
```

Comment:
See [here](https://github.com/openfoodfacts/taxonomy-editor/pull/438/files/eb939ba8758378cf5cee103e69ee4a11baf59d79#r1525339709) as to why I return `None` instead of throwing exceptions 
This was the initial implementation I wrote: the name search terms had to be surrounded by quotes if you used a semi-colon
But I realised that it was bothersome to write the en:dairy query for example or any kind of language:search_value query.
In general, to add terms to the search bar, users rarely think about adding quotes so I wanted quotes to not be necessary to search text
Moreover, it allows us to reserve quotes for exact full text search

Reply:
huuum ok I see you point after writing my previous rants ;-)

it's true that we already have columns all over the place.

I still think we can make your implementation a bit more organized in this case…:
* skip term if it's quoted or as no :
* split on ":", see if it's a recognized keyword or skip
* handle filter